kubeadm: Add 'kubeadm upgrade plan' and 'kubeadm upgrade apply' CLI commands

luxas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

This PR is splitted out from: https://github.com/kubernetes/kubernetes/pull/48899 and only handles the CLI/command code. It adds no-op functions only to `phases/upgrade`.

A large chunk of this code is unit tests.
The code here should be pretty straightforward as there is no actual upgrade or business logic here.
It would be cool to get this merged soon-ish.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes: https://github.com/kubernetes/kubeadm/issues/14

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```
@kubernetes/sig-cluster-lifecycle-pr-reviews PTAL

pipejakob commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+	upgrade.PrepullImagesInParallel(prepuller, flags.pullTimeout)
+
+	// Now; perform the upgrade procedure
+	if err := PerformUpgrade(flags, client, internalcfg); err != nil {
```
Based on the implementation, maybe this should be `UpgradeControlPlaneImages()` or something more specific? It was unexpected that `PerformUpgrade()` wasn't actually the full upgrade, and that taking care of things like RBAC/ServiceAccounts/etc. were considered "post-upgrade tasks," which I expected to just be client-side cleanup.

luxas replied:
I think `UpgradeControlPlaneComponents()` might be the best describing name here, thanks for the inspiration.

luxas replied:
fixed

pipejakob commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+// ValidateFlags makes sure the arguments given to apply are valid
+func ValidateFlags(flags *applyFlags) error {
+	if len(flags.newK8sVersionStr) == 0 {
+		return fmt.Errorf("--version is a required parameter. Please specify the version you want to upgrade to")
```
My personal preference, although I hope others weigh in on the expected UX here: the version to upgrade to is so central to the operation of upgrading (and is required) that I wouldn't even expect it to be a flag. I would actually expect this to just be a required argument, so you would do:

    $ kubeadm upgrade 1.8.0

instead of:

    $ kubeadm upgrade --version 1.8.0

To me, it's a little like having to type `ping --ip-address 127.0.0.1`.

luxas replied:
@lukemarsden WDYT?

(This is just copied from what we said some time in March :smile:, happy to revisit)

lukemarsden replied:
Given that the version is optional (if you don't specify it, it'll upgrade to the latest stable right?) I think a flag is the right thing to do.

lukemarsden replied:
Oh, if it's required, then I agree with @pipejakob

luxas replied:
it is always mandatory, will change to an arg instead

luxas replied:
now changed to `kubeadm upgrade apply [version]` instead.

pipejakob commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+
+	// Specify the valid flags specific for apply
+	cmd.Flags().BoolVarP(&flags.nonInteractiveMode, "yes", "y", flags.nonInteractiveMode, "Perform the upgrade and do not prompt for confirmation (non-interactive mode)")
+	cmd.Flags().BoolVarP(&flags.force, "force", "f", flags.force, "Force upgrading although some requirements might not be met. This also implies non-interactive mode")
```
UI nitpick: If we're using a period in the middle of the string to end a sentence, it feels bare without a period at the end as well.

luxas replied:
fixed

pipejakob commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid due to these fatal errors: %v.\n", versionSkewErrs.Mandatory)
```
It seems like an antipattern to both print an error *and* return it. Shouldn't we just return a well-formatted error so that whoever chooses to handle it can print the appropriate message? Otherwise, we're asking for duplicates.

luxas replied:
fixed

pipejakob reviewed (commented):
Still making my way through the diff, but here are some initial comments:

pipejakob commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,62 @@
+
+// StateBeforeUpgrade describes the current state before any upgrade is made
+type StateBeforeUpgrade struct {
+	KubeVersion, DNSVersion, KubeadmVersion string
```
Nitpick: I think this sort of field compaction hurts readability.

luxas replied:
fixed

pipejakob commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,62 @@
+// Upgrade defines an upgrade possibility to upgrade from a current version to a new one
+type Upgrade struct {
+	Description string
+	Before      StateBeforeUpgrade
```
Little weird to me that we have both `StateBeforeUpgrade` and `StateAfterUpgrade` as separate types, even though they're nearly identical. Were you worried about type safety? I would have expected to see:

    Before  ClusterState
    After   ClusterState

Since one of them has a single KubeletVersion and the other has a list, we could just make it a list of one element?
&lt;/bikeshedding&gt;

luxas replied:
> we could just make it a list of one element

I'll do that. Basically, this was to make the output look like this:
```
COMPONENT   CURRENT      AVAILABLE
Kubelet     1 x v1.7.0   v1.7.3
            1 x v1.7.1   v1.7.3
            1 x v1.7.2   v1.7.3
```

(happy to revisit that part as well)

luxas replied:
fixed

pipejakob commented on cmd/kubeadm/app/phases/upgrade/versiongetter.go:
```diff
@@ -0,0 +1,79 @@
+// versionGetter defines an interface for fetching different versions.
+// Easy to implement a fake variant of this interface for unit testing
+type versionGetter interface {
+	GetClusterVersion() (string, *versionutil.Version, error)
```
Same comments as before about the unnecessary `Get` prefix:

https://golang.org/doc/effective_go.html#Getters

luxas replied:
will do

luxas replied:
fixed

pipejakob reviewed (commented):
Few more comments:

timothysc commented:
@luxas thx for breaking this up, it makes it much easier to walk through.  I'll take a look in a bit.

luxas commented:
@pipejakob Did something break when you said `Few more comments:`, but no comments from are present after that?

@timothysc Thanks, my pleasure. Depending on what you say I here now I'll break up the rest of the code as much as it goes by `one-file-per-PR` after this.

pipejakob commented:
@luxas Yup, no idea why, but it looks like the new comments got merged into the old review and the new review was empty? Doesn't look like it dropped any comments, though.

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+func NewCmdApply(parentFlags *cmdUpgradeFlags) *cobra.Command {
+	flags := &applyFlags{
+		parent:      parentFlags,
+		pullTimeout: 15 * time.Minute,
```
`SetImplicitFlags` SGTM

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
```
Nitpick: >500 LoC here are net-new unit tests for this PR, which is what contributes heavily to the size.
I can see tomorrow if there is any easy-ish way for me to split that up better

luxas replied:
removed the daemonset prepulling stub code for now, will add later

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+		return err
+	}
+
+	// If interactive mode is set (--yes was not specified), ask ther user whether they really want to upgrade
```
fixed

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+
+	// If interactive mode is set (--yes was not specified), ask ther user whether they really want to upgrade
+	// If --yes is specified; just proceed and do it
+	if !flags.nonInteractiveMode {
```
added helper func `flags.SessionIsInteractive()` for better readability

luxas commented:
@pipejakob @timothysc Ready for re-review. Fixed @pipejakob's comments and cleaned up some things; removed the daemonset prepulling logic for now, etc.

lukemarsden commented:
Great work, thank you @luxas!

Modulo the UX feedback I've given [here](https://docs.google.com/document/d/10hKRks1QEBvmM-wYNGGel-QNcdJa5ATKoQdauytFt2o/edit), this LGTM! ✨

pipejakob commented:
I'll take a look at this today.

pipejakob commented:
/lgtm

We should still discuss if version should have a flag or just be a toplevel argument, but definitely not blocking right now.

luxas commented:
I'll let @timothysc take a look at this tonight and then fixup the verify failure and re-lgtm tomorrow.
Thanks @pipejakob for reviewing, yeah, let's take that UX question in tomorrow's kubeadm meeting :+1:

timothysc commented on cmd/kubeadm/app/phases/upgrade/configuration.go:
```diff
@@ -0,0 +1,57 @@
+		return fmt.Errorf("couldn't read from standard input: %v", err)
+	}
+	answer := scanner.Text()
+	if strings.ToLower(answer) != "y" {
```
You should also check 'yes' here b/c people are bad at paying attention to prompts ;-)

luxas replied:
fixed

timothysc commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	// Run healthchecks against the cluster
+	if err := upgrade.VerifyClusterHealth(client); err != nil {
```
Might want to chance the signature name to something more binary if you're only returning error. 
IsClusterAlive...

luxas replied:
`CheckClusterHealth` (it's more than /healthz only...)

luxas replied:
set to `CheckClusterHealth` now

timothysc commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	// Fetch the configuration from a file or ConfigMap and validate it
+	cfg, err := upgrade.FetchConfiguration(client, os.Stdout, cfgPath, nonInteractively)
```
why is fetching the config pass in the 'Interactive' variable.

luxas replied:
I had it to be; kubeadm will ask the user if it should proceed with the default config if it didn't exist in-cluster. Now we can require every v1.7 user to do `kubeadm config upload from-flags`

luxas replied:
fixed

timothysc commented on cmd/kubeadm/app/cmd/upgrade/common_test.go:
```diff
@@ -0,0 +1,124 @@
+				KubernetesVersion: "v1.7.1",
+			},
+			expectedBytes: []byte(`[upgrade/config] Configuration used:
+	api:
```
This seems pretty brittle vs in case you change things, couldn't you just regex parse the output for the key strings that matter?

luxas replied:
I think this makes sense, as it is very explicit with the end state

timothysc commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -0,0 +1,137 @@
+		fmt.Fprintf(w, "Upgrade to the latest %s:\n", upgrade.Description)
+		fmt.Fprintln(w, "")
+		fmt.Fprintln(tabw, "COMPONENT\tCURRENT\tAVAILABLE")
+		fmt.Fprintf(tabw, "API Server\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
```
Might want to consider https://github.com/ryanuber/columnize here, or do similar wrapping.

luxas replied:
not for these PRs, but maybe in the future. This works pretty well and is stdlib. Not wanna vendor sth now.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/plan_test.go:
```diff
@@ -0,0 +1,328 @@
+				},
+			},
+			expectedBytes: []byte(`Components that must be upgraded manually after you've upgraded the control plane with 'kubeadm upgrade apply':
+COMPONENT   CURRENT      AVAILABLE
```
Same comment re trimming own the output string to check here.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/upgrade.go:
```diff
@@ -0,0 +1,75 @@
+// or with an invalid subcommand, cobra will print usage information, but still exit cleanly.
+// We want to return an error code in these cases so that the
+// user knows that their command was invalid.
+func subCmdRunE(name string) func(*cobra.Command, []string) error {
```
Isn't this a crib, why not make it more globally accessible.

luxas replied:
tech debt; should find a place for it...

timothysc commented on cmd/kubeadm/app/cmd/upgrade/upgrade.go:
```diff
@@ -0,0 +1,75 @@
+
+	cmd.PersistentFlags().StringVar(&flags.kubeConfigPath, "kubeconfig", flags.kubeConfigPath, "The KubeConfig file to use for talking to the cluster.")
+	cmd.PersistentFlags().StringVar(&flags.cfgPath, "config", flags.cfgPath, "Path to kubeadm config file (WARNING: Usage of a configuration file is experimental).")
+	cmd.PersistentFlags().BoolVar(&flags.allowExperimentalUpgrades, "allow-experimental-upgrades", flags.allowExperimentalUpgrades, "Show unstable versions of Kubernetes as an upgrade alternative and allow upgrading to an alpha/beta/release candidate versions of Kubernetes.")
```
Is anyone other then us going to care about the rc/experimental distinction?  

rc's are experimental.

luxas replied:
@roberthbailey asked for an explicit flag for this and I kind of agree (we agreed in the meeting as well, I think you were there)

timothysc commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -0,0 +1,63 @@
+		return true
+	}
+	// Don't report something available for upgrade if we don't know the current state
+	if len(u.Before.KubeletVersions) == 0 {
```
I'm flummoxed by this logic.

luxas replied:
ideas how to improve the kubelet version handling?

luxas commented on cmd/kubeadm/app/phases/upgrade/policy.go:
```diff
@@ -0,0 +1,33 @@
+
+// EnforceVersionPolicies enforces that the proposed new version is compatible with all the different version skew policies
+func EnforceVersionPolicies(_ VersionGetter, _ string, _ *version.Version, _, _ bool) *VersionSkewPolicyErrors {
+	// No-op now and return no skew errors
```
yes

luxas commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// upgradeVariables holds variables needed for performing an upgrade or planning to do so
+type upgradeVariables struct {
```
naming suggestion?

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+	cmd.Flags().BoolVarP(&flags.force, "force", "f", flags.force, "Force upgrading although some requirements might not be met. This also implies non-interactive mode.")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied. This also implies non-interactive mode.")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to.")
+	cmd.Flags().DurationVar(&flags.pullTimeout, "pull-timeout", flags.pullTimeout, "The maximum amount of time to wait for the control plane pods to be downloaded.")
```
kind of agree, but not for this PR

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+
+// applyFlags holds the information about the flags that can be passed to apply
+type applyFlags struct {
+	nonInteractiveMode bool
```
unfortunately `requireUserApproval` != `nonInteractiveMode`

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+	if len(flags.newK8sVersionStr) == 0 {
+		return fmt.Errorf("--version is a required parameter. Please specify the version you want to upgrade to")
+	}
+	var err error
```
it was because of the assignment directly to cfg; now being explicit and using a temporary variable

luxas commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -0,0 +1,192 @@
+	// Specify the valid flags specific for apply
+	cmd.Flags().BoolVarP(&flags.nonInteractiveMode, "yes", "y", flags.nonInteractiveMode, "Perform the upgrade and do not prompt for confirmation (non-interactive mode).")
+	cmd.Flags().BoolVarP(&flags.force, "force", "f", flags.force, "Force upgrading although some requirements might not be met. This also implies non-interactive mode.")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied. This also implies non-interactive mode.")
```
fixed.

timothysc reviewed (commented):
/lgtm

rmmh commented:
/test pull-kubernetes-bazel

luxas committed:
kubeadm: Resolve tech debt; move commonly used funcs to a general package instead of duplicating

luxas committed:
kubeadm: Add 'kubeadm upgrade plan' and 'kubeadm upgrade apply' CLI commands

luxas committed:
Add unit tests for kubeadm upgrade|plan

luxas commented:
Simple rebase

luxas commented:
/retest

luxas commented:
FYI @ixdy @spxtr the verify job timed out here... Known issue?

luxas committed:
autogenerated bazel

luxas commented:
/retest
